From 7cc1eb3546a20a09ce8a7313744225dceaae53d8 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Fri, 3 Jul 2020 12:37:37 +0800
Subject: [PATCH 94/98] backend-drm: Support dynamic control of the compositor

Use a config file to control the compositor's state.

The default config file is "/tmp/.weston_drm.conf". This can be
overridden using the "WESTON_DRM_CONFIG" environment variable.

Supported config formats are "compositor:<key>:<value>". For example:
echo "compositor:state:sleep" > /tmp/.weston_drm.conf
echo "compositor:state:block" > /tmp/.weston_drm.conf
echo "compositor:state:freeze" > /tmp/.weston_drm.conf
echo "compositor:state:offscreen" > /tmp/.weston_drm.conf
echo "compositor:state:off" > /tmp/.weston_drm.conf
echo "compositor:state:on" > /tmp/.weston_drm.conf
echo "compositor:hotplug:off" > /tmp/.weston_drm.conf
echo "compositor:hotplug:on" > /tmp/.weston_drm.conf
echo "compositor:hotplug:force" > /tmp/.weston_drm.conf
echo "compositor:cursor:hide" > /tmp/.weston_drm.conf
echo "compositor:cursor:show" > /tmp/.weston_drm.conf
echo "compositor:output:pin" > /tmp/.weston_drm.conf
echo "compositor:output:unpin" > /tmp/.weston_drm.conf
echo "compositor:output:refresh" > /tmp/.weston_drm.conf

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 libweston/backend-drm/drm.c | 75 ++++++++++++++++++++++++++++++++++++-
 libweston/compositor.c      |  8 +++-
 libweston/libinput-seat.c   |  4 ++
 3 files changed, 84 insertions(+), 3 deletions(-)

diff --git a/libweston/backend-drm/drm.c b/libweston/backend-drm/drm.c
index 9723d8c0..80061f66 100644
--- a/libweston/backend-drm/drm.c
+++ b/libweston/backend-drm/drm.c
@@ -4955,6 +4955,75 @@ config_handle_output(struct drm_backend *b, const char *name,
 	}
 }
 
+static void
+config_handle_compositor(struct drm_backend *b, const char *key,
+			 const char *value)
+{
+	if (!strcmp(key, "state")) {
+		if (!strcmp(value, "sleep")) {
+			weston_compositor_sleep(b->compositor);
+		} else if (!strcmp(value, "block")) {
+			udev_input_disable(&b->input);
+		} else if (!strcmp(value, "freeze")) {
+			udev_input_disable(&b->input);
+			weston_compositor_offscreen(b->compositor);
+		} else if (!strcmp(value, "offscreen")) {
+			/* HACK: Enter offscreen mode and turn off DPMS. */
+			weston_compositor_sleep(b->compositor);
+			weston_compositor_offscreen(b->compositor);
+			udev_input_enable(&b->input);
+		} else if (!strcmp(value, "off")) {
+			udev_input_disable(&b->input);
+			weston_compositor_sleep(b->compositor);
+		} else {
+			/* HACK: Force wake from offscreen state. */
+			if (b->compositor->state == WESTON_COMPOSITOR_OFFSCREEN)
+				b->compositor->state = WESTON_COMPOSITOR_IDLE;
+
+			weston_compositor_wake(b->compositor);
+			weston_compositor_damage_all(b->compositor);
+			udev_input_enable(&b->input);
+		}
+	} else if (!strcmp(key, "hotplug")) {
+		if (!strcmp(value, "off"))
+			wl_event_source_fd_update(b->udev_drm_source, 0);
+		else if (!strcmp(value, "on"))
+			wl_event_source_fd_update(b->udev_drm_source,
+						  WL_EVENT_READABLE);
+		else if (!strcmp(value, "force"))
+			hotplug_update_handler(b);
+	} else if (!strcmp(key, "cursor")) {
+		if (!strcmp(value, "hide"))
+			b->compositor->hide_cursor = true;
+		else if (!strcmp(value, "show"))
+			b->compositor->hide_cursor = false;
+
+		b->compositor->view_list_needs_rebuild = true;
+		weston_compositor_damage_all(b->compositor);
+	} else if (!strcmp(key, "mirror")) {
+		/* Set output mirror mode (off/stretch/fit) */
+		if (!strcmp(value, "off"))
+			b->compositor->output_mirror =
+				WESTON_OUTPUT_MIRROR_NONE;
+		else if (!strcmp(value, "stretch"))
+			b->compositor->output_mirror =
+				WESTON_OUTPUT_MIRROR_STRETCH;
+		else if (!strcmp(value, "fit") || !strcmp(value, "on"))
+			b->compositor->output_mirror =
+				WESTON_OUTPUT_MIRROR_FIT;
+		weston_compositor_schedule_heads_changed(b->compositor);
+	} else if (!strcmp(key, "output")) {
+		if (!strcmp(value, "pin"))
+			b->compositor->pin_output = true;
+		else if (!strcmp(value, "unpin"))
+			b->compositor->pin_output = false;
+		else if (!strcmp(value, "refresh"))
+			drm_device_recovery_required(b->drm);
+
+		weston_compositor_damage_all(b->compositor);
+	}
+}
+
 static int
 config_timer_handler(void *data)
 {
@@ -4996,7 +5065,9 @@ config_timer_handler(void *data)
 
 	/**
 	 * Parse configs, formatted as <type>:<key>:<value>
-	 * For example: "output:all:rotate90"
+	 * For example:
+	 *	output:all:rotate90
+	 *	compositor:state:off
 	 */
 	while (3 == fscanf(conf_fp,
 			   "%" STR(MAX_CONF_LEN) "[^:]:"
@@ -5004,6 +5075,8 @@ config_timer_handler(void *data)
 			   "%" STR(MAX_CONF_LEN) "[^\n]%*c", type, key, value)) {
 		if (!strcmp(type, "output"))
 			config_handle_output(b, key, value);
+		else if (!strcmp(type, "compositor"))
+			config_handle_compositor(b, key, value);
 	}
 
 	fclose(conf_fp);
diff --git a/libweston/compositor.c b/libweston/compositor.c
index 812dd3c3..c8764993 100644
--- a/libweston/compositor.c
+++ b/libweston/compositor.c
@@ -6249,6 +6249,10 @@ weston_compositor_wake(struct weston_compositor *compositor)
 {
 	uint32_t old_state = compositor->state;
 
+	/* HACK: Prevent waking up from offscreen state via input events. */
+	if (old_state == WESTON_COMPOSITOR_OFFSCREEN)
+		return;
+
 	/* The state needs to be changed before emitting the wake
 	 * signal because that may try to schedule a repaint which
 	 * will not work if the compositor is still sleeping */
@@ -6256,10 +6260,10 @@ weston_compositor_wake(struct weston_compositor *compositor)
 
 	switch (old_state) {
 	case WESTON_COMPOSITOR_SLEEPING:
+		wl_signal_emit(&compositor->wake_signal, compositor);
+		/* fall through */
 	case WESTON_COMPOSITOR_IDLE:
-	case WESTON_COMPOSITOR_OFFSCREEN:
 		weston_compositor_dpms(compositor, WESTON_DPMS_ON);
-		wl_signal_emit(&compositor->wake_signal, compositor);
 		/* fall through */
 	default:
 		wl_event_source_timer_update(compositor->idle_source,
diff --git a/libweston/libinput-seat.c b/libweston/libinput-seat.c
index 8aa50449..d5f2377e 100644
--- a/libweston/libinput-seat.c
+++ b/libweston/libinput-seat.c
@@ -328,6 +328,10 @@ udev_input_enable(struct udev_input *input)
 	struct udev_seat *seat;
 	int devices_found = 0;
 
+	/* Already enabled */
+	if (input->libinput_source && !input->suspended)
+		return 0;
+
 	loop = wl_display_get_event_loop(c->wl_display);
 	fd = libinput_get_fd(input->libinput);
 	input->libinput_source =
-- 
2.39.5

